risc0/zkvm/src/recursion/receipt.rs

```diff
@@ -0,0 +1,288 @@
+        tagged_struct("risc0.SystemState", &[self.image_id], &[self.pc])
+    }
+}
+
+impl ReceiptMeta {
+    /// todo
+    pub fn decode(flat: &mut VecDeque<u32>) -> Self {
+        Self {
+            input: read_sha_halfs(flat),
+            pre: SystemState::decode(flat),
+            post: SystemState::decode(flat),
+            sys_exit: flat.pop_front().unwrap() as u8,
+            user_exit: flat.pop_front().unwrap() as u8,
+            output: read_sha_halfs(flat),
+        }
+    }
+    /// todo
+    pub fn encode(&self, flat: &mut Vec<u32>) {
+        write_sha_halfs(flat, &self.input);
+        self.pre.encode(flat);
+        self.post.encode(flat);
+        flat.push(self.sys_exit as u32);
+        flat.push(self.user_exit as u32);
+        write_sha_halfs(flat, &self.output);
+    }
+    /// todo
+    pub fn digest(&self) -> Digest {
+        tagged_struct(
+            "risc0.ReceiptMeta",
+            &[
+                self.input,
+                self.pre.digest(),
+                self.post.digest(),
+                self.output,
+            ],
+            &[self.sys_exit.into(), self.user_exit.into()],
+        )
+    }
+}
+
+/// todo
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SegmentRecursionReceipt {
+    /// todo
+    pub seal: Vec<u32>,
+    /// todo
+    pub control_id: Digest,
+    /// todo
+    pub meta: ReceiptMeta,
+}
+
+impl SegmentRecursionReceipt {
+    /// todo
+    #[cfg(not(target_os = "zkvm"))]
+    pub fn verify(&self) -> Result<(), VerificationError> {
+        use risc0_core::field::baby_bear::BabyBearElem;
+        use risc0_zkp::core::hash::poseidon::PoseidonHashSuite;
+
+        use super::CircuitImpl;
+
+        // Make the hal
+        let hal = risc0_zkp::verify::CpuVerifyHal::<_, PoseidonHashSuite, _>::new(&CIRCUIT_CORE);
+        let valid_ids = valid_control_ids();
+        let check_code = |_po2: u32, control_id: &Digest| -> Result<(), VerificationError> {
+            let Some(_) = valid_ids.iter().position(|elem| elem == control_id) else {
+                return Err(VerificationError::JournalDigestMismatch);
+            };
+            Ok(())
+        };
+        // Verify the receipt itself is correct
+        risc0_zkp::verify::verify(&hal, &CIRCUIT_CORE, &self.seal, check_code)?;
+        // Extract the globals from the seal
+        let output_elems: &[BabyBearElem] =
+            bytemuck::cast_slice(&self.seal[..CircuitImpl::OUTPUT_SIZE]);
+        let mut seal_meta = VecDeque::<u32>::new();
+        for elem in output_elems {
+            seal_meta.push_back(elem.as_u32())
+        }
+        // TODO: Read root hash
+        seal_meta.drain(0..16);
+        // Verify the output hash matches that data
+        let output_hash = read_sha_halfs(&mut seal_meta);
+        if output_hash != self.meta.digest() {
+            return Err(VerificationError::JournalDigestMismatch);
+        }
+        // Everything passed
+        Ok(())
+    }
+}
+
+/// todo
```

Comment:
This is the datatype bonsai will most likely return, so we should probably get this fully documented.